archive/extract.go

```diff
@@ -15,10 +16,35 @@ type PathMode struct {
 	Mode os.FileMode
 }
 
+var (
+	umaskLock      sync.Mutex
+	extractCounter int
+	originalUmask  int
+)
```

Comment:
We can put these three variables inside a private struct if anyone thinks it'll be clearer.
Or we can even move the variables and the related functions (`setUmaskIfNeeded` and `unsetUmaskIfNeeded`) to an internal package.

Personally, I don't think it's necessary but it's an option.